src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,283 @@
+        {
+            if (fixes.IsEmpty
+                || !document.TryGetText(out var text))
+            {
+                return document;
+            }
+
+            var codeActions = GetCodeActionsForFixes(document, fixes);
+            var getChangesTasks = codeActions.Select(action => GetChangesForCodeActionAsync(document, action, cancellationToken));
+
+            await Task.WhenAll(getChangesTasks).ConfigureAwait(false);
+
+            var allTextChanges = Enumerable.Empty<TextChange>();
+            var allAddedProjectReferences = Enumerable.Empty<ProjectReference>();
+            var allAddedMetaDataReferences = Enumerable.Empty<MetadataReference>();
+
+            foreach (var getChangesTask in getChangesTasks)
+            {
+                var (projectChanges, textChanges) = await getChangesTask.ConfigureAwait(false);
+
+                allTextChanges = allTextChanges.Concat(textChanges);
+                allAddedProjectReferences = allAddedProjectReferences.Concat(projectChanges.GetAddedProjectReferences());
+                allAddedMetaDataReferences = allAddedMetaDataReferences.Concat(projectChanges.GetAddedMetadataReferences());
+            }
+
+            var newProject = document.Project.AddMetadataReferences(allAddedMetaDataReferences);
+            newProject = newProject.AddProjectReferences(allAddedProjectReferences);
+
+            var newText = text.WithChanges(allTextChanges);
+            var newDocument = newProject.GetDocument(document.Id).WithText(newText);
+
+            return newDocument;
+        }
+
+        private ImmutableArray<CodeAction> GetCodeActionsForFixes(Document document, ImmutableArray<AddImportFixData> fixes)
+        {
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+            var codeActions = ArrayBuilder<CodeAction>.GetInstance();
+
+            foreach (var fix in fixes)
+            {
+                if (addImportService.TryCreateCodeAction(document, fix, out var codeAction))
+                {
+                    codeActions.Add(codeAction);
+                }
+            }
+
+            return codeActions.ToImmutableAndFree();
+        }
+
+        private async Task<(ProjectChanges projectChanges, IEnumerable<TextChange> textChanges)> GetChangesForCodeActionAsync(Document document, CodeAction codeAction, CancellationToken cancellationToken)
+        {
+            var newSolution = await codeAction.GetChangedSolutionAsync(new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+            var newDocument = newSolution.GetDocument(document.Id);
+
+            var textChanges = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var projectChanges = newDocument.Project.GetChanges(document.Project);
+
+            return (projectChanges, textChanges);
+        }
+
+        public bool Equals(AddImportFixData x, AddImportFixData y)
```

Comment:
why is Equals/GetHashCode necessary?

Reply:
It was so I could get a distinct set of fixes in `GetUnambiguousFixesAsync` although I could apply a distinct set of TextChanges in `ApplyFixesAsync` instead,